### ISSUE TITLE
Bug: [padding-line-between-statements] function overload should not padded with blankline

### ISSUE THREAD

**Issue by Shinigami92** [typescript-eslint/typescript-eslint#6291]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=4.9.3&sourceType=module&code=C4TwDgpgBAwg9gWwRAdsAIgSwM5gDYCGIACgQE7BQC8UBKIA3AFBMQAeYcFUAZgK4oAxsExwUUAFZxMKACoEA5qQrYAFGHLBsALliJkaLLkIlNAbQC6ASl3ZgZGQubtO3fkJFjJ0uYuVb1TWwAfl14JFQMHHwif0sbKDsHFAUoAB8oAQATCB4ZCCznDi5Kd2FRcSkZeSUgwJVQvQjDaJM461t7R3TMlBy8lAKoAG8mKCgyCGA+MnENBoA6BAIwVXrgK2oAPih54AXgdg2FqpRVAHJzq2YAXxYAegAqR7HHqAAlCAQ4ADcIbCggkQYEweGgLkm2GwFQBwAAFgRKFkxOdKAQsllaPQoD8CHg+NAeFwJhB0RAyNgFq9XlAAALzAgIHF4glQWRw6D9Ah8PCUXH46DAOCAsF0KlQN60ybTWYA9nQQSiwaYrk8vksiDix73JhlTziRWklDKRlTcnoXLc3mqfkEzrJBQJJKOZh6ioio0m5CHMgWnhW4A2jWNZ0pJ1dFI9bK5fKFXUCcpeQ10L1m32WtVBgUhiOO+3dDLRgZDUbjTA8KBZ1lUGgoHl4Tal8YkmXiIux5jjO5jKD3e5QADKXTAUFAkEBBDsOigACJzkS4AAjcjnWgAgQAaxQcAA7uJJ2yZ9tZ-O4EuVzOe9KZuJbZrJjFBBBVP2D5vt3u11AzABBAC0ABaBB-gAXgAPFsFgANT3AANFAlzXEwdxAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6YgQwBMvmBzWoyaJaAI0T4A7okRNayfB3yIAtrPyoMAbXDYciaNAD20SABpde4FFHwOTANYAZZonRQO8SR0KozUYmhEADd3SAAqcyhhAA98MNsjMgdBSgc3MABfSzAAXV1szKA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA

### Repro Code

```typescript
export function joinTagParts(parts: CommentDisplayPart[]): string;
export function joinTagParts(parts?: CommentDisplayPart[]): string | undefined;
export function joinTagParts(parts?: CommentDisplayPart[]): string | undefined {
  return parts?.map((part) => part.text).join('');
}

/**
 * Removes compile expressions that don't add any value for readers.
 *
 * @param value The default value to clean.
 * @returns The cleaned default value.
 */
function cleanParameterDefault(value: string): string;
function cleanParameterDefault(value?: string): string | undefined;
function cleanParameterDefault(value?: string): string | undefined {
  if (value == null) {
    return undefined;
  }

  // Strip type casts: "'foobar' as unknown as T" => "'foobar'"
  return value.replace(/ as unknown as [A-Za-z<>]+/, '');
}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/<rule-name>": ["error", { blankLine: 'always', prev: '*', next: 'block-like' }],
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    // ...
  }
}
```

### Expected Result

```ts
export function joinTagParts(parts: CommentDisplayPart[]): string;
export function joinTagParts(parts?: CommentDisplayPart[]): string | undefined;
export function joinTagParts(parts?: CommentDisplayPart[]): string | undefined {
  return parts?.map((part) => part.text).join('');
}

/**
 * Removes compile expressions that don't add any value for readers.
 *
 * @param value The default value to clean.
 * @returns The cleaned default value.
 */
function cleanParameterDefault(value: string): string;
function cleanParameterDefault(value?: string): string | undefined;
function cleanParameterDefault(value?: string): string | undefined {
  if (value == null) {
    return undefined;
  }

  // Strip type casts: "'foobar' as unknown as T" => "'foobar'"
  return value.replace(/ as unknown as [A-Za-z<>]+/, '');
}
```

### Actual Result

```ts
export function joinTagParts(parts: CommentDisplayPart[]): string;
export function joinTagParts(parts?: CommentDisplayPart[]): string | undefined;

export function joinTagParts(parts?: CommentDisplayPart[]): string | undefined {
  return parts?.map((part) => part.text).join('');
}

/**
 * Removes compile expressions that don't add any value for readers.
 *
 * @param value The default value to clean.
 * @returns The cleaned default value.
 */
function cleanParameterDefault(value: string): string;
function cleanParameterDefault(value?: string): string | undefined;

function cleanParameterDefault(value?: string): string | undefined {
  if (value == null) {
    return undefined;
  }

  // Strip type casts: "'foobar' as unknown as T" => "'foobar'"
  return value.replace(/ as unknown as [A-Za-z<>]+/, '');
}
```

### Additional Info

Maybe there is a regression as this was reported and fixed in a previous version if I understand correctly:

- https://github.com/typescript-eslint/typescript-eslint/pull/4345
- https://github.com/typescript-eslint/typescript-eslint/issues/4030

**Comment by ericbf** [comment]
>Seems like an edge case that wasn't handled before. Which is unsurprising, given how annoyingly complex these formatting rules are. Which is part of why [we strongly recommend not using them](https://typescript-eslint.io/linting/troubleshooting/formatting). 😄

@JoshuaKGoldberg the problem is that prettier specifically [has opted out of formatting blank lines](https://prettier.io/docs/en/rationale.html#empty-lines), choosing to just preserve them as they are in the source file, save removing multiple subsequent lines. We’re left with either using this rule, or integrating yet another formatter to get blank lines consistent in the codebase 😢

**Comment by bradzacher** [comment]
My HOT🌶️ take - the existence of one single blank line between statements is not a problem you should care about as it does not meaningfully increase the readability of your code, and you should not use this lint rule at all for that reason.

There are cases where you will want a blank line to separate things and there are cases where you do not want a blank line to help group things (eg between overloads or tightly related functions) - and it's impossible to encode that nuance in a lint rule.

**Comment by bradzacher** [comment]
Hi there 👋!

As you may have heard [ESLint core is deprecating its formatting rules](https://eslint.org/blog/2023/10/deprecating-formatting-rules/). We are likewise moving in the same direction and deprecating our formatting rules.

Deprecated rules will be removed in the next major version as their new home is the [eslint-stylistic](https://github.com/eslint-stylistic/eslint-stylistic) project.

If this issue is important to you please consider filling an issue over at eslint-stylistic which will be the new home for stylistic/formatting rules.

Thanks!
